mpeg2/src/interleaving_muxer.rs

```diff
@@ -0,0 +1,189 @@
+                self.streams[stream_index].last_written_ts = t;
+            }
+        } else {
+            // and then we need to see if this enabled any other streams to emit their outputs
+            self.streams[stream_index].buffered_packets.push_back(p);
+            if let Some(t) = ts {
+                self.largest_ts_in_buffer = self.largest_ts_in_buffer.max(t);
+            }
+            self.write_packets_outside_buffer_duration()?;
+
+            // find a stream with smallest packet timestamp and emit all packets in this stream if packets
+            // in all other streams with timestamps are larger
+            let mut packets = vec![];
+            let mut min_ts_stream_index = 0;
+            let mut last_written_ts = 0;
+            if let Some((index, _)) = self
+                .streams
+                .iter()
+                .enumerate()
+                .filter(|(index, _)| *index != stream_index)
+                .min_by_key(|(_, s)| {
+                    if let Some(p) = s.buffered_packets.front() {
+                        p.dts_90khz.or(p.pts_90khz).unwrap_or(s.last_written_ts)
+                    } else {
+                        u64::MAX
+                    }
+                })
+            {
+                let min_ts = self.min_ts_excluding(self.streams[index].wrapper.packet_id() as usize);
+                let stream = &mut self.streams[index];
+                min_ts_stream_index = index;
+                while let Some(p) = stream.buffered_packets.front() {
+                    let ts = p.dts_90khz.or(p.pts_90khz).unwrap_or(stream.last_written_ts);
+                    if ts <= min_ts {
+                        packets.push(stream.buffered_packets.pop_front().unwrap());
+                        last_written_ts = ts;
+                    } else {
+                        break;
+                    }
+                }
+            }
+            if !packets.is_empty() {
+                for packet in packets {
+                    self.write_packet(min_ts_stream_index, packet)?;
+                }
+                self.streams[min_ts_stream_index].last_written_ts = last_written_ts;
+            }
+        }
+        Ok(())
+    }
+
+    fn write_packet(&mut self, stream_index: usize, p: Packet) -> Result<(), EncodeError> {
+        let stream = &mut self.streams[stream_index].wrapper;
+        self.muxer.write(stream, p)?;
+        Ok(())
+    }
+
+    fn write_packets_outside_buffer_duration(&mut self) -> Result<(), EncodeError> {
+        let mut packets = vec![];
+        for (stream_index, stream) in self.streams.iter_mut().enumerate() {
+            while let Some(p) = stream.buffered_packets.front() {
+                let ts = p.dts_90khz.or(p.pts_90khz).unwrap_or(stream.last_written_ts);
+                if self.largest_ts_in_buffer - ts > self.max_buffer_duration {
+                    packets.push((stream_index, stream.buffered_packets.pop_front().unwrap()));
```

Comment:
Why not just call `self.write_packet` here?

Also, prefer `expect` over `unwrap`:

```suggestion
                    packets.push((stream_index, stream.buffered_packets.pop_front().expect("there must be at least one packet to pop")));
```

Reply:
seeing the following errors when calling `self.write_packet` directly, I couldn't find a way to walk around.
```bash
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> mpeg2/src/interleaving_muxer.rs:114:21
    |
109 |         for (stream_index, stream) in self.streams.iter_mut().enumerate() {
    |                                       -----------------------------------
    |                                       |
    |                                       first mutable borrow occurs here
    |                                       first borrow later used here
...
114 |                     self.write_packet(stream_index, stream.buffered_packets.pop_front().unwrap());
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
```